addons/etcd/scripts/common.sh

```diff
@@ -0,0 +1,111 @@
+#!/bin/sh
+
+# config file used to bootstrap the etcd cluster
+configFile=$TMP_CONFIG_PATH
+
+checkBackupFile() {
+  local backupFile=$1
+  output=$(etcdutl snapshot status ${backupFile})
+  # check if the command was successful
+  if [ $? -ne 0 ]; then
+    echo "ERROR: Failed to check the backup file with etcdutl"
+    exit 1
+  fi
+  # extract the total key from the output
+  totalKey=$(echo $output | awk -F', ' '{print $3}')
+  # check if total key is a number
+  case $totalKey in
+    *[!0-9]*)
+      echo "ERROR: snapshot totalKey is not a valid number."
+      exit 1
+      ;;
+  esac
+
+  # define a threshold to check if the total key count is too low
+  # consider increasing this value when dealing with larger etcd clusters
+  threshold=$BACKUP_KEY_THRESHOLD #[modifiable]
+  if [ "$totalKey" -lt $threshold ]; then
+    echo "WARNING: snapshot totalKey is less than the threshold"
+    exit 1
+  fi
+}
+
+getClientProtocol() {
+  # check client tls if is enabled
+  line=$(grep 'advertise-client-urls' ${configFile})
+  if echo $line | grep -q 'https'; then
+    echo "https"
+  elif echo $line | grep -q 'http'; then
+    echo "http"
+  fi
+}
+
+getPeerProtocol() {
+  # check peer tls if is enabled
+  line=$(grep 'initial-advertise-peer-urls' ${configFile})
+  if echo $line | grep -q 'https'; then
+    echo "https"
+  elif echo $line | grep -q 'http'; then
+    echo "http"
+  fi
+}
+
+execEtcdctl() {
+  local endpoints=$1
+  shift
+  clientProtocol=$(getClientProtocol)
+  tlsDir=$TLS_MOUNT_PATH
+  # check if the clientProtocol is https and the tlsDir is not empty
+  if [ $clientProtocol = "https" ] && [ -d "$tlsDir" ] && [ -s "${tlsDir}/ca.crt" ] && [ -s "${tlsDir}/tls.crt" ] && [ -s "${tlsDir}/tls.key" ]; then
+    etcdctl --endpoints=${endpoints} --cacert=${tlsDir}/ca.crt --cert=${tlsDir}/tls.crt --key=${tlsDir}/tls.key "$@"
+  elif [ $clientProtocol = "http" ]; then
+    etcdctl --endpoints=${endpoints} "$@"
+  else
+    echo "ERROR: bad etcdctl args: clientProtocol:${clientProtocol}, endpoints:${endpoints}, tlsDir:${tlsDir}, please check!"
+    exit 1
+  fi
+  # check if the etcdctl command was successful
+  if [ $? -ne 0 ]; then
+    echo "etcdctl command failed"
+    exit 1
+  fi
+}
+
+# this function will be deprecated in the future
+execEtcdctlNoCheckTLS() {
+  local endpoints=$1
+  shift
+  etcdctl --endpoints=${endpoints} "$@"
+  # check if the etcdctl command was successful
+  if [ $? -ne 0 ]; then
+    echo "etcdctl command failed"
+    exit 1
+  fi
+}
+
+updateLeaderIfNeeded() {
+  local retries=$1
+
+  if [ $retries -le 0 ]; then
+    echo "Maximum number of retries reached, leader is not ready"
+    exit 1
+  fi
+
+  status=$(execEtcdctlNoCheckTLS ${leaderEndpoint} endpoint status)
```

Comment:
why use execEtcdctlNoCheckTLS instead of execEtcdctl  in updateLeaderIfNeeded

Reply:
does it work if the protocol is https?

Reply:
It doesn't work for https, cause updateLeaderIfNeeded only used by switchover action. Which cannot mount data volume and tls volume, so it's unable to check tls right now

It's an temporary solution and will be deprecated after Implementing volume mount for switchover

related issue #7875

Reply:
so if https is used, switchover doesn't work right now, right?

Reply:
yes